File: backend/src/main/java/com/appbackend/backend/entity/Petition.java
package com.appbackend.backend.entity;

import java.time.LocalDateTime;

import com.appbackend.backend.enums.PetitionStatus;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.PrePersist;
import jakarta.persistence.Table;
import lombok.Data;

@Entity
@Data
@Table(name = "petitions")
public class Petition {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String title;

    @Column(nullable = false)
    private String receiver;

    @Column(name = "image_url")
    private String imageUrl;
    private String imagePublicId;

    @Column(nullable = false, columnDefinition = "TEXT")
    private String problem;

    @Column(nullable = false, columnDefinition = "TEXT")
    private String solution;
    
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id")
    private User user;

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(nullable = false)
    private int votes = 0;

    @Column(name = "official_response")
    private String officialResponse;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private PetitionStatus status = PetitionStatus.ACTIVE;

    @PrePersist
    void prePersist() {
        if (createdAt == null) {
            createdAt = LocalDateTime.now();
        }

        if (status == null) {
            status = PetitionStatus.ACTIVE;
        }
    }
}

File: frontend/src/Types/petition.ts
export type PetitionStatus = "ACTIVE" | "CLOSED" | "BANNED";

export interface Petition {
  id: number;
  title: string;
  receiver: string;
  imageUrl?: string | null;
  imagePublicId?: string | null;
  problem: string;
  solution: string;
  status?: PetitionStatus | string;
  createdAt?: string;
  officialResponse?: string | null;
  votes: number;
  userId?: number | null;
  userFullName?: string | null;
  userContact?: string | null;
}

export interface PetitionCreateRequest {
  title: string;
  receiver: string;
  problem: string;
  solution: string;
  userId?: number;
}

export interface PetitionVoteRequest {
  signerName: string;
}

